File: backend/app/schemas/application.py
"""지원 트래커 API 스키마."""

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.models.application import (
    ApplicationStatus,
    EmploymentType,
    EndStage,
    HiringType,
)


# ---------------------------------------------------------------- 회사 · 트랙 · 문서
class CompanyCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    industry: str | None = Field(default=None, max_length=100)
    size: str | None = Field(default=None, max_length=50)
    website: str | None = Field(default=None, max_length=500)
    memo: str | None = None

    @field_validator("name")
    @classmethod
    def _strip(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("회사명을 입력해주세요")
        return v


class CompanyUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    industry: str | None = None
    size: str | None = None
    website: str | None = None
    memo: str | None = None


class CompanyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    industry: str | None = None
    size: str | None = None
    website: str | None = None
    memo: str | None = None
    application_count: int = 0


class CompanyRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str


class TrackCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    color: str | None = Field(default=None, pattern=r"^#[0-9a-fA-F]{6}$")
    sort_order: int = 0


class TrackUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    color: str | None = Field(default=None, pattern=r"^#[0-9a-fA-F]{6}$")
    sort_order: int | None = None


class TrackOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    color: str | None = None
    sort_order: int = 0


DOC_TYPE_PATTERN = r"^(resume|cover_letter|portfolio|experience|other)$"


class DocumentCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    doc_type: str = Field(default="resume", pattern=DOC_TYPE_PATTERN)
    content: str | None = None


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    title: str
    doc_type: str
    created_at: datetime
    updated_at: datetime


class DocumentRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    title: str
    doc_type: str = "resume"
    doc_type_label: str | None = None


class DocumentInput(BaseModel):
    """지원에 붙일 제출물 하나. id 가 오면 소유권 확인, title 이 오면 get-or-create."""

    id: UUID | None = None
    title: str | None = Field(default=None, max_length=200)
    doc_type: str = Field(default="resume", pattern=DOC_TYPE_PATTERN)

    @model_validator(mode="after")
    def _need_one(self) -> "DocumentInput":
        if not self.id and not (self.title and self.title.strip()):
            raise ValueError("제출물은 id 나 제목 중 하나가 필요해요")
        return self


# ---------------------------------------------------------------- 지원
class ApplicationCreate(BaseModel):
    """company_id 또는 company_name(get-or-create) 중 하나는 필수. track/document 도 같은 방식."""

    company_id: UUID | None = None
    company_name: str | None = Field(default=None, max_length=200)
    track_id: UUID | None = None
    track_name: str | None = Field(default=None, max_length=100)
    resume_document_id: UUID | None = None
    resume_document_title: str | None = Field(default=None, max_length=200)
    # 이력서 외 제출물(포트폴리오·경험기술서 등). 주면 제출물 집합 전체를 이걸로 맞춘다.
    documents: list[DocumentInput] | None = None

    position: str = Field(min_length=1, max_length=200)
    title: str | None = Field(default=None, max_length=300)  # 비면 "{회사} - {포지션}"
    posting_url: str | None = Field(default=None, max_length=1000)
    status: ApplicationStatus = ApplicationStatus.DISCOVERED
    end_stage: EndStage | None = None
    employment_type: EmploymentType | None = None
    hiring_type: HiringType | None = None
    discovered_at: date | None = None
    applied_at: date | None = None
    deadline_at: date | None = None
    next_event_at: datetime | None = None
    next_action: str | None = Field(default=None, max_length=500)
    retrospective: str | None = None

    @model_validator(mode="after")
    def _company_required(self) -> "ApplicationCreate":
        if not self.company_id and not (self.company_name and self.company_name.strip()):
            raise ValueError("회사를 선택하거나 회사명을 입력해주세요")
        return self


class ApplicationUpdate(BaseModel):
    """상태(status/end_stage)는 여기서 못 바꾼다 — PATCH /{id}/status 로 강제."""

    company_id: UUID | None = None
    company_name: str | None = Field(default=None, max_length=200)
    track_id: UUID | None = None
    track_name: str | None = Field(default=None, max_length=100)
    clear_track: bool = False
    resume_document_id: UUID | None = None
    resume_document_title: str | None = Field(default=None, max_length=200)
    clear_resume_document: bool = False
    documents: list[DocumentInput] | None = None  # 주면 제출물 집합 전체를 대체
    clear_documents: bool = False

    position: str | None = Field(default=None, min_length=1, max_length=200)
    title: str | None = Field(default=None, max_length=300)
    posting_url: str | None = Field(default=None, max_length=1000)
    employment_type: EmploymentType | None = None
    hiring_type: HiringType | None = None
    discovered_at: date | None = None
    applied_at: date | None = None
    deadline_at: date | None = None
    next_event_at: datetime | None = None
    next_action: str | None = Field(default=None, max_length=500)
    retrospective: str | None = None


class StatusChange(BaseModel):
    status: ApplicationStatus
    end_stage: EndStage | None = None
    note: str | None = Field(default=None, max_length=500)


class HistoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    from_status: str | None
    to_status: str
    changed_at: datetime
    note: str | None = None
    source: str


class ApplicationOut(BaseModel):
    id: UUID
    title: str
    position: str
    posting_url: str | None
    status: str
    status_label: str
    end_stage: str | None
    end_stage_label: str | None
    employment_type: str | None
    employment_label: str | None
    hiring_type: str | None
    hiring_label: str | None
    discovered_at: date | None
    applied_at: date | None
    deadline_at: date | None
    next_event_at: datetime | None
    next_action: str | None
    retrospective: str | None
    created_at: datetime
    updated_at: datetime
    # 파생
    d_day: int | None
    season: str | None
    is_active: bool
    is_applied: bool
    passed_docs: bool
    reached_interview: bool
    company: CompanyRef
    track: TrackOut | None
    documents: list[DocumentRef]
    resume_document: DocumentRef | None  # documents 중 이력서 종류의 첫 번째(파생)


class ApplicationDetailOut(ApplicationOut):
    history: list[HistoryOut]


class ApplicationListOut(BaseModel):
    items: list[ApplicationOut]
    total: int


class StatusChangeOut(BaseModel):
    application: ApplicationOut
    celebration: bool


class SeasonStat(BaseModel):
    season: str
    total: int
    applied: int
    passed_docs: int
    interview: int
    offer: int


class TrackStat(BaseModel):
    track_id: UUID | None
    track_name: str
    total: int
    applied: int
    passed_docs: int
    interview: int
    offer: int


class DocumentStat(BaseModel):
    """제출물 버전별 승률 — 자소서가 빠진 전형에서 '무엇을 냈나'를 가르는 축."""

    document_id: UUID
    title: str
    doc_type: str
    doc_type_label: str
    total: int
    applied: int
    passed_docs: int
    interview: int
    offer: int


class StatsOut(BaseModel):
    by_status: dict[str, int]
    by_season: list[SeasonStat]
    by_track: list[TrackStat]
    by_document: list[DocumentStat]
    funnel: dict[str, int]  # applied / passed_docs / interview / offer
    active_count: int
    total: int


# ---------------------------------------------------------------- 임포트
class ImportRowError(BaseModel):
    row: int
    field: str | None
    reason: str
    raw: str | None = None


class ImportReport(BaseModel):
    dry_run: bool
    total_rows: int
    created: int
    updated: int
    skipped: int
    errors: list[ImportRowError]
    preview: list[dict]

File: frontend/src/types/application.ts
// 지원 트래커 타입 — backend/app/models/application.py 의 라벨 dict 를 미러한다. 키가 바뀌면 양쪽을 함께 고칠 것.

export type ApplicationStatus =
  | "discovered" | "reviewing" | "applied" | "doc_passed" | "coding_test" | "assignment" | "interview"
  | "offer" | "rejected" | "withdrawn" | "no_response" | "not_applied";

export const STATUS_LABELS: Record<ApplicationStatus, string> = {
  discovered: "발견", reviewing: "검토중", applied: "지원완료", doc_passed: "서류통과",
  coding_test: "코테·필기", assignment: "과제", interview: "면접", offer: "최종합격",
  rejected: "탈락", withdrawn: "포기", no_response: "무응답", not_applied: "미지원",
};

/** 🔥 진행중 보드 컬럼 순서 */
export const ACTIVE_STATUSES: readonly ApplicationStatus[] = [
  "discovered", "reviewing", "applied", "doc_passed", "coding_test", "assignment", "interview",
];
export const CLOSED_STATUSES: readonly ApplicationStatus[] = ["offer", "rejected", "withdrawn", "no_response", "not_applied"];
export const CLOSED_WITH_STAGE: readonly ApplicationStatus[] = ["rejected", "withdrawn"];

export const STATUS_COLORS: Record<ApplicationStatus, string> = {
  discovered: "#8a94a6", reviewing: "#8a94a6", applied: "#5b9bd5", doc_passed: "#6dae8a",
  coding_test: "#e5c07b", assignment: "#e5c07b", interview: "#f2a66d", offer: "#e06c75",
  rejected: "#6b7280", withdrawn: "#8b6f5a", no_response: "#c98fb8", not_applied: "#6b7280",
};

export type EndStage = "document" | "coding_test" | "written_test" | "assignment" | "interview" | "final";
export const END_STAGE_LABELS: Record<EndStage, string> = {
  document: "서류", coding_test: "코테", written_test: "필기", assignment: "과제", interview: "면접", final: "최종",
};

export type EmploymentType = "full_time" | "contract" | "intern_conversion" | "intern_experience" | "bootcamp";
export const EMPLOYMENT_LABELS: Record<EmploymentType, string> = {
  full_time: "정규직", contract: "계약직", intern_conversion: "채용연계형 인턴",
  intern_experience: "체험형 인턴", bootcamp: "부트캠프/교육",
};

export type HiringType = "open_recruitment" | "rolling" | "always_open";
export const HIRING_LABELS: Record<HiringType, string> = { open_recruitment: "공채", rolling: "수시", always_open: "상시" };

export interface CompanyRef { id: string; name: string }
export interface Company extends CompanyRef {
  industry: string | null; size: string | null; website: string | null; memo: string | null; application_count: number;
}
export interface Track { id: string; name: string; color: string | null; sort_order: number }
export type DocType = "resume" | "cover_letter" | "portfolio" | "experience" | "other";
/** 자소서를 빼고 이력서·포트폴리오·경험기술서를 함께 받는 전형이 늘어 종류를 넓혔다 */
export const DOC_TYPE_LABELS: Record<DocType, string> = {
  resume: "이력서", cover_letter: "자기소개서", portfolio: "포트폴리오",
  experience: "경험기술서", other: "기타",
};
export interface DocumentRef { id: string; title: string; doc_type: DocType; doc_type_label?: string | null }
export interface DocumentItem extends DocumentRef { created_at: string; updated_at: string }
/** 지원에 붙일 제출물 — id 나 title 중 하나 */
export interface DocumentInput { id?: string; title?: string; doc_type?: DocType }

export interface Application {
  id: string;
  title: string;
  position: string;
  posting_url: string | null;
  status: ApplicationStatus;
  status_label: string;
  end_stage: EndStage | null;
  end_stage_label: string | null;
  employment_type: EmploymentType | null;
  employment_label: string | null;
  hiring_type: HiringType | null;
  hiring_label: string | null;
  discovered_at: string | null;   // YYYY-MM-DD
  applied_at: string | null;
  deadline_at: string | null;
  next_event_at: string | null;   // ISO datetime
  next_action: string | null;
  retrospective: string | null;
  created_at: string;
  updated_at: string;
  d_day: number | null;
  season: string | null;
  is_active: boolean;
  is_applied: boolean;
  passed_docs: boolean;
  reached_interview: boolean;
  company: CompanyRef;
  track: Track | null;
  documents: DocumentRef[];
  resume_document: DocumentRef | null;   // documents 중 이력서 종류의 첫 번째(파생)
}

export interface HistoryEntry {
  id: string; from_status: ApplicationStatus | null; to_status: ApplicationStatus;
  changed_at: string; note: string | null; source: "user" | "agent" | "import";
}
export interface ApplicationDetail extends Application { history: HistoryEntry[] }

export interface ApplicationInput {
  company_id?: string; company_name?: string;
  track_id?: string; track_name?: string; clear_track?: boolean;
  resume_document_id?: string; resume_document_title?: string; clear_resume_document?: boolean;
  documents?: DocumentInput[]; clear_documents?: boolean;   // documents 를 주면 제출물 집합 전체를 대체
  position?: string; title?: string; posting_url?: string | null;
  status?: ApplicationStatus; end_stage?: EndStage | null;
  employment_type?: EmploymentType | null; hiring_type?: HiringType | null;
  discovered_at?: string | null; applied_at?: string | null; deadline_at?: string | null;
  next_event_at?: string | null; next_action?: string | null; retrospective?: string | null;
}

export interface Stats {
  by_status: Partial<Record<ApplicationStatus, number>>;
  by_season: { season: string; total: number; applied: number; passed_docs: number; interview: number; offer: number }[];
  by_track: { track_id: string | null; track_name: string; total: number; applied: number; passed_docs: number; interview: number; offer: number }[];
  by_document: { document_id: string; title: string; doc_type: DocType; doc_type_label: string; total: number; applied: number; passed_docs: number; interview: number; offer: number }[];
  funnel: { applied: number; passed_docs: number; interview: number; offer: number };
  active_count: number;
  total: number;
}

export interface ImportReport {
  dry_run: boolean; total_rows: number; created: number; updated: number; skipped: number;
  errors: { row: number; field: string | null; reason: string; raw: string | null }[];
  preview: Record<string, unknown>[];
}

/** 프런트에서도 D-day 를 다시 계산한다 — 자정을 넘기면 서버 값이 낡는다 */
export function dDay(deadline: string | null): number | null {
  if (!deadline) return null;
  const [y, m, d] = deadline.split("-").map(Number);
  if (!y || !m || !d) return null;
  const today = new Date();
  const a = Date.UTC(y, m - 1, d);
  const b = Date.UTC(today.getFullYear(), today.getMonth(), today.getDate());
  return Math.round((a - b) / 86400000);
}

export function ddayLabel(n: number | null): string {
  if (n === null) return "";
  if (n === 0) return "D-Day";
  return n > 0 ? `D-${n}` : `D+${-n}`;
}
